walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,63 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+
 
 public final class FindMeetingQuery {
+  /**Return all possible time intervals for a meeting, taking into account the optional attendees, 
+   if it's possible, otherwise return all the possible time intervals without them.*/
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> meetingsWithOptional = findOptionalTimeForMeeting(events, request, true);
+    if (!meetingsWithOptional.isEmpty()) {
+        return meetingsWithOptional;
+    }
+    return findOptionalTimeForMeeting(events, request, false);
+  }
+
+  /**Return all possible time intervals for a meeting, according to its attendees' events.*/
+  private ArrayList<TimeRange> findOptionalTimeForMeeting(Collection<Event> events, MeetingRequest request, boolean withOptional) {
```

Comment:
Google best practices recommend to avoid boolean parameters as much as possible; instead, we recommend to use two separate methods, whose names reflect the effect of the boolean, and possibly sharing aspects of the implementation. (The reasons to avoid boolean parameters are (a) they're hard to extend, and (b) they more-or-less guarantee higher complexity - more "if"s - within the method).

In this case, you could even have a single method "findTimeForMeeting" that accepts a "Collection<String> attendees", and two helpers - "mandatoryAttendees(MeetingRequest)" and "allAttendees(MeetingRequest)".